### ISSUE TITLE
Support for bundled multiple client operations

### ISSUE THREAD

**Issue by pburkholder** [chef/chef-rfc#124]
There are some use cases that I don't know how to tackle, so consider this issue a proposal for an RFC, instead of an RFC itself:

> As a Chef user provisioning a node <br>
> when a chef client run is terminated midway through it's run, between the 'client create' and 'node save' steps <br>
> I don't want ghost client objects in the chef server that aren't associated with node objects

> As a Chef user making updates to Chef Vault <br>
> when knife vault rotate-keys is executed <br>
> I don't want my vault left unusable because foo/bar_keys updated but foo/bar did not [0]

The individual REST API calls to ChefServer are atomic, but sometimes the user of ChefServer needs multiple operations to succeed, as in the above examples, to maintain some consistent state about the infrastructure.

One thing that occurs to me is to create a /write-ahead endpoint where a client could make a request for several calls, e.g.:

    PUT /writeahead:
    {
      actions: [
        { endpoint: /client, body: {client_body}, method: "PUT" },
        { endpoint: /node, body: {node_body}, method: "PUT"}
      ]
    }

The server would then attempt to execute all the actions in the request, and give a 200 OK only after all of them succeed. If the client were to crash or otherwise disappear, then at least the overall state of the server is still something one can reason about.

I've not attempted a full RFC at this point because:

- I don't know if this is at all useful in the community (asides from a few customers I've spoken with)
- I don't know if this utter crazytalk
- I do know that any suggestion along these lines would need to be completely rewritten, so I thought I'd get some feedback before I put together a proposal going in the wrong direction.

REFERENCES

\[0\]: https://github.com/Nordstrom/chef-vault/blob/master/THEORY.md

**Comment by ranjib** [comment]
have you looked at the node desired state API? that should help with some of the challenges.
https://github.com/chef/chef-rfc/blob/master/rfc045-node_state_separation.md

**Comment by pburkholder** [comment]
@ranjib I have, but it doesn't seem like it would cover the case of autoscaling nodes which fail to create by their node or client object, or the data bags case.

**Comment by coderanger** [comment]
This is hugely out of scope for Chef Server. Building a transactional API would require rebuidling basically everything from the ground up. Related to the "ghost client" issue, the new validator-less bootstraps do not have this problem as both the client and node objects are created by knife before starting the bootstrap. This doesn't help with autoscaling in particular as that is a case when you are likely to use old-style bootstrapping, but a better fix would probably be to make the validator-based bootstraps work more like the new style.

**Comment by pburkholder** [comment]
@coderanger I agree 100% about transactions. But isn't there something in between that would help with those edges cases that become common at scale?

**Comment by coderanger** [comment]
@pburkholder All of these are examples that your tooling is broken, basically. For example, the bootstrap system just kind of lazily handled node creation as a side effect of the first run. We've outgrown that for a variety of reasons, and the new bootstrap model fixes those issues. As for data bags, they are [pretty bad and  you shouldn't use them](https://coderanger.net/data-bags/). The bags API is riddled with those kinds of issues because it is a semantic-free database. If you need a better secrets management system, [I have a post on that too](https://coderanger.net/chef-secrets/) though it is a bit out of date given new systems like Sneaker, KeyWhiz, and Hashicorp Vault (at varying levels of bleeding edge).

**Comment by ranjib** [comment]
mitigating broken registration (client/node is registered but not converged) is a common requirement. My understanding is node state API will allow you catch those issues. What to do with them is up to the tool/service/process that is triggering the provisioning, as registration can be done on the box using validation certs of from outside. Reaping zombie nodes (as i call them :-) ) should be part of the autoscaling tool/process ( or other provisioning mechanism, like chef provisioning, knife ec2 create etc).

**Comment by coderanger** [comment]
You can do it today too, the bootstrap code should just save the initial run data (including run list) before the converge starts in the same way as the validator-less bootstraps work. I just think no one has had the cycles to go back and fix the old code. It also needs to be fixed to generate the client key locally if it doesn't already, but that is unrelated. The deeper problem is "failed initial converge" when that first run is being driven from something like cloud-init via autoscaling. Figuring out how to react to that kind of failure is complex, mostly due to how little feedback AWS (and other) autoscaling transmits back up to the control plane.

**Comment by danielsdeleo** [comment]
@coderanger yeah, so saving the node at the start of the run is something we used to do, and could do quite easily. The difficulty there is you break people using search-on-run-list-item as service discovery; for example, a new web server can be added to rotation before it's configured (and that's even worse if chef-client fails for some reason).

**Comment by coderanger** [comment]
@danielsdeleo It would have to be before any cookbook loading or run list stuffs happen, so `node['roles']` et al wouldn't be populated.

**Comment by danielsdeleo** [comment]
That's exactly what we do now, I think. But it leaves you with a node with an empty run list if the initial chef-client run fails.

**Comment by coderanger** [comment]
So save it after setting run_list and chef_environment (and any `-j` attribute data) but before expanding/processing?

**Comment by danielsdeleo** [comment]
I'm pretty sure any change along those lines will break a few people, since people search on the run list as well. Though if there is something more optimal and we can make it opt-in until the next major rev, that'd be cool.

**Comment by coderanger** [comment]
@danielsdeleo Pretty sure what I'm describing is exactly how validatorless bootstraps work already, so no difference. If there are people that would be broken, it already happened. https://github.com/chef/chef/blob/master/lib/chef/knife/bootstrap/client_builder.rb#L125-L145

**Comment by danielsdeleo** [comment]
Right, validatorless is opt-in, so you can just stop using it until your search usage is "fixed"  (or avoid altogether).

**Comment by coderanger** [comment]
Sure, this could be an option on chef-client in the same way :-) If the target is specifically auto-scaling machines (as the primary situation where validatorless can't be used) that would be fine, just build it in to the image.

**Comment by coderanger** [comment]
Also while validatorless is technically opt-in, it is in a confusing way. A way more likely reason that the validator key file wouldn't be found is that the user didn't know to download it or otherwise screwed up their config. I wouldn't call that an opt-in to any useful degree, lots of people have probably upgraded accidentally.

**Comment by pburkholder** [comment]
@coderanger @danielsdeleo @ranjib:
 
Thanks all for your comments. I pitched this idea on behalf of a customer who's had ghost client issues in their particular autoscaling setup, and noted that chef-vault/data-bags had a similar need for more reliability around some API operations that do need to happen in sequence.

I'm pretty much convinced, indeed, that chef-vault and databags are going to have reliability issues, and one should look for other data backends if one needs transactional guarantees.

As for client/node, I could use a little more unpacking around a few statements since I'm not as familiar with recent API developments. To wit:

@coderanger: "If the target is specifically auto-scaling machines (as the primary situation where validatorless can't be used) that would be fine, just build it in to the image."  -- I'm not sure what 'build it into the image' means here since I'm unsure about the antecedent to 'it', can you clarify?

@ranjib: "mitigating broken registration (client/node is registered but not converged) is a common requirement. My understanding is node state API will allow you catch those issues." -- Can you expound on that? I'm not seeing how the new API there helps when the client is created but not the node. (I do see how it helps when the client/node is created but not converged, since we can differentiate between desired state and the 'nil' current state.)

@coderanger: "For example, the bootstrap system just kind of lazily handled node creation as a side effect of the first run. We've outgrown that for a variety of reasons, and the new bootstrap model fixes those issues."  -- I believe you've aleady addressed that in your comments with @danielsdeleo, but if you mean something different I'd love to know.

I'm looking forward to closing this issue out, and I hope the discussion here pretty well placates anyone who does ask about Chef and transactional guarantees.

Thanks again, Peter

**Comment by lamont-granquist** [comment]
I still don't understand what the major problem is with 'ghost' node/clients or how transactions will fix the problem.

If 'ghost' nodes are causing provisioning to break, then the code in the validatorless bootstraps fixes this via the option to delete nodes and clients if they exist.  It should not fail there.

The validatorless bootstraps also do not require an admin key they require a user which has perms to delete and create nodes/clients -- this can be a specially created user for this purpose.

When provisioning the server itself fails because EC2 craps out or something goes wrong with the virt before chef-client has converged the node then there's really nothing much you can do about that.  You will get ghost clients and nodes.  If you are doing search operations in order to determine virts to send traffic to then you need to be able to search not only on the run_list of the virt but also an attribute which indicates that it has been provisioned.  Simply adding a 'provisioned' tag to the node in the recipe and allowing the final node.save to update that field and then filtering search results on that tag would be enough.  Then the presence of ghost nodes should not be an operational concern and a gargage collection task can periodically run to clean up nodes which have been seen by the GC job for awhile and have not transitioned to the provisioned state.

So questions:

So which 'ghost node' issue are we trying to solve first, and how did they get there?  Because I don't see how transactional guarantees are supposed to solve that problem.  Are you supposed to wrap the entire provisioning of the node with a transaction?  That doesn't make any sense, while just adding some kind of searchable attribute that tags a node as being fully provisioned does.

And the approach of 'transactions' seems flawed from the outset.  It is trying to create a perfect database of imperfect servers.  Servers will crash and lockup and go into swap and perform badly and there has to be a 'liveness' criteria associated with the nodes in the database in order for them to be useful at all.  You can't expect to just dump the node database and get a list of useful servers without some kind of filtering.  Trying to make the database perfect seems flawed.  Adding state changes to the database, on the other hand, can be done atomically and can add value and can be done trivially today.

This also complies with promise theory as well.  The node in this case is responsible for announcing to the world "i have been provisioned" via the final node.save -- the server is not made responsible for ensuring that the provisioning transaction to the node has completed.   

So, where is the requirement for transaction guarantees?  What is the process which they are following where they think they need them?  Because while "transactional guarantees" sounds good and like something we 'should' support, it is vastly simpler to engineer a reliable system that doesn't need them.  It also looks like we have the primitives available to be able to engineer a reliable system now without having them.  Thus, they'd just be adding complex possibly buggy code and APIs to a system that did not need them.  Where's the concrete problem that we can't solve without transactions?

**Comment by coderanger** [comment]
Data bags will always have issues even with pseudo-transaction system because they don't support partial updates. Two people running a get-update-put cycle will still have race conditions in a gross way. If you are hitting these issues, it is a good sign that you should stop using data bags and get a real database/API that knows your semantics.

The ghost node thing is mostly an annoyance if the server comes up correctly but the recipes fail part way through. With the new bootstrap mode, that's fine because if it got far enough to run the chef-client recipe, the service will be installed and on the next interval run it will try again. With old-style bootstraps, the run list and other data will be lost forever because the node is only saved at the end. Adding the same kind of node save to the startup process during bootstrap (maybe via `--bootstrap` on the command line) would make things a bit more robust for minimal downside.

As for what I mean by "bake it in", that new command line argument would have to be added either in your AMI or in your cloud-init userdata, whichever actually launches chef-client. Regardless, this would make it opt-in, preferably with a deprecation for the old way so we can make it the default in 13.

**Comment by pburkholder** [comment]
Closing this as it's clearly the wrong answer to the wrong question.
